libs/vgc/ui/margins.h

```diff
@@ -0,0 +1,305 @@
+// Copyright 2022 The VGC Developers
+// See the COPYRIGHT file at the top-level directory of this distribution
+// and at https://github.com/vgc/vgc/blob/master/COPYRIGHT
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef VGC_UI_MARGINS_H
+#define VGC_UI_MARGINS_H
+
+#include <vgc/geometry/rect2f.h>
+#include <vgc/geometry/vec4f.h>
+#include <vgc/ui/api.h>
+
+namespace vgc::ui {
+
+/// \class vgc::ui::Margins
+/// \brief Represents a set of 4 margins for the 4 sides of a UI element.
+///
+/// Note that margins are allowed to be negative, in which case adding a margin
+/// to a rect would shrink the rect.
+///
+// VGC_UI_API <- Omitted on purpose.
+//               If needed, manually export individual functions.
+class Margins {
+private:
+    enum Indices_ {
+        Top = 0,
+        Right,
+        Bottom,
+        Left
+    };
+
+public:
+    /// Constructs a `Margins` with all margins set to 0.
+    ///
+    constexpr Margins()
+        : v_() {
+    }
+
+    /// Constructs a `Margins` with all margins set to the given `margin`.
+    ///
+    constexpr Margins(float margin)
```

Comment:
should this be explicit?

Reply:
We decided to keep it explicit. There seems to be no ambiguity possibly, and the implicit conversion can be convenient.